openshift/openshift_transport.go

```diff
@@ -44,67 +43,32 @@ func (t openshiftTransport) ValidatePolicyConfigurationScope(scope string) error
 	namespace       string
 	stream          string
-	tag             string
-	dockerReference reference.Named // Computed from the above in advance, so that later references can not fail.
+	dockerReference reference.NamedTagged // Computed from the above in advance, so that later references can not fail.
 }
 
-// FIXME: Is imageName like this a good way to refer to OpenShift images?
-// Keep this in sync with scopeRegexp!
-var imageNameRegexp = regexp.MustCompile("^([^:/]*)/([^:/]*):([^:/]*)$")
-
 // ParseReference converts a string, which should not start with the ImageTransport.Name prefix, into an OpenShift ImageReference.
-func ParseReference(reference string) (types.ImageReference, error) {
-	// Overall, this is modelled on openshift/origin/pkg/cmd/util/clientcmd.New().ClientConfig() and openshift/origin/pkg/client.
-	cmdConfig := defaultClientConfig()
-	logrus.Debugf("cmdConfig: %#v", cmdConfig)
-	restConfig, err := cmdConfig.ClientConfig()
-	if err != nil {
-		return nil, err
-	}
-	// REMOVED: SetOpenShiftDefaults (values are not overridable in config files, so hard-coded these defaults.)
-	logrus.Debugf("restConfig: %#v", restConfig)
-	baseURL, _, err := restClientFor(restConfig)
-	if err != nil {
-		return nil, err
-	}
-	logrus.Debugf("URL: %#v", *baseURL)
-
-	m := imageNameRegexp.FindStringSubmatch(reference)
-	if m == nil || len(m) != 4 {
-		return nil, fmt.Errorf("Invalid image reference %s, %#v", reference, m)
+func ParseReference(ref string) (types.ImageReference, error) {
+	r, err := reference.ParseNamed(ref)
+	// I believe it's an hard requirement to be namespace/stream:tag right?
+	tagged, ok := r.(reference.NamedTagged)
+	if err != nil && !ok {
+		return nil, fmt.Errorf("Invalid image reference %s, %#v", ref, r)
 	}
 
-	return NewReference(baseURL, m[1], m[2], m[3])
+	return NewReference(tagged)
 }
 
 // NewReference returns an OpenShift reference for a base URL, namespace, stream and tag.
-func NewReference(baseURL *url.URL, namespace, stream, tag string) (types.ImageReference, error) {
-	// Precompute also dockerReference so that later references can not fail.
-	//
-	// This discards ref.baseURL.Path, which is unexpected for a “base URL”;
-	// but openshiftClient.doRequest actually completely overrides url.Path
-	// (and defaultServerURL rejects non-trivial Path values), so it is OK for
-	// us to ignore it as well.
-	//
-	// FIXME: This is, strictly speaking, a namespace conflict with images placed in a Docker registry running on the same host.
-	// Do we need to do something else, perhaps disambiguate (port number?) or namespace Docker and OpenShift separately?
-	dockerRef, err := reference.WithName(fmt.Sprintf("%s/%s/%s", baseURL.Host, namespace, stream))
-	if err != nil {
-		return nil, err
-	}
-	dockerRef, err = reference.WithTag(dockerRef, tag)
-	if err != nil {
-		return nil, err
+func NewReference(dockerRef reference.NamedTagged) (types.ImageReference, error) {
+	r := strings.SplitN(dockerRef.RemoteName(), "/", 2)
```

Comment:
Shouldn’t this be …`, 3)` to detect and refuse too nested references unsupported by OpenShift?